### PR TITLE
ast: recover from popping too many modes

### LINKED ISSUE #2603: Whitespace before a build instruction causes the parser to panic from an empty stack.
Whitespace before a build instruction causes the parser to panic from an empty stack.

AFAIK leading whitespace shouldn't be unparseable (please correct me if I'm wrong); but if so, we should return a more appropriate error message, rather than panicking and throwing out a full backtrace.

`earthly version v0.6.30 743687991d227f2f7e2e6ba11a90f443b4f99c67 linux/amd64; Ubuntu 20.04.5 LTS (Focal Fossa)`

#### Okay:
```Earthfile
VERSION --use-cache-command 0.6

## Build Arguments ###########################################################

ARG SourceImage=ubuntu:focal
#^^^
 
RUN true
```
---
```shell
 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Found buildkit daemon as podman container (earthly-buildkitd)

 2. Build 🔧
————————————————————————————————————————————————————————————————————————————————

Share your logs with an Earthly account (experimental)! Register for one at https://ci.earthly.dev.
Error: build target: build main: failed to solve: Earthfile line 0:0 target build not found
in              +build
```

### Crashes:
```Earthfile
VERSION --use-cache-command 0.6

## Build Arguments ###########################################################

 ARG SourceImage=ubuntu:focal
#^^^
 
RUN true
```
---
```shell
 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Found buildkit daemon as podman container (earthly-buildkitd)

 2. Build 🔧
————————————————————————————————————————————————————————————————————————————————

panic:  [recovered]
        panic: Empty Stack [recovered]
        panic: Empty Stack [recovered]
        panic: Empty Stack

goroutine 168 [running]:
github.com/earthly/earthly/ast/parser.(*EarthParser).EarthFile.func2()
        /earthly/ast/parser/earth_parser.go:722 +0xfe
panic({0x1069f20, 0x1434ab0})
        /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/earthly/earthly/ast/parser.(*EarthParser).Stmts.func2()
        /earthly/ast/parser/earth_parser.go:1855 +0xfe
panic({0x1069f20, 0x1434ab0})
        /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/antlr/antlr4/runtime/Go/antlr.(*BaseLexer).PopMode(0xc0003d2980?)
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/lexer.go:267 +0x129
github.com/earthly/earthly/ast.(*lexer).popRecipeMode(0xc0003d2980)
        /earthly/ast/lexer.go:58 +0x9d
github.com/earthly/earthly/ast.(*lexer).processIndentation(0xc0003d2980, {0x144d0e0, 0xc00033de80})
        /earthly/ast/lexer.go:160 +0x293
github.com/earthly/earthly/ast.(*lexer).NextToken(0xc0003d2980)
        /earthly/ast/lexer.go:104 +0x1c5
github.com/antlr/antlr4/runtime/Go/antlr.(*CommonTokenStream).fetch(0xc0000eb280, 0x3e8)
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/common_token_stream.go:128 +0x69
github.com/antlr/antlr4/runtime/Go/antlr.(*CommonTokenStream).Fill(0xc0000eb280)
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/common_token_stream.go:370 +0x3d
github.com/antlr/antlr4/runtime/Go/antlr.(*CommonTokenStream).GetTextFromInterval(0xc0000eb280, 0xc000823728)
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/common_token_stream.go:334 +0x3b
github.com/antlr/antlr4/runtime/Go/antlr.(*CommonTokenStream).GetTextFromTokens(0x417c6d?, {0x144d0e0?, 0xc00033d7d0?}, {0x144d0e0, 0xc00033d7d0})
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/common_token_stream.go:325 +0x87
github.com/antlr/antlr4/runtime/Go/antlr.(*DefaultErrorStrategy).ReportNoViableAlternative(0x2?, {0x1457800, 0xc0006cb5b0}, 0xc000a453b0)
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/error_strategy.go:258 +0xa2
github.com/antlr/antlr4/runtime/Go/antlr.(*DefaultErrorStrategy).ReportError(0x0?, {0x1457800?, 0xc0006cb5b0?}, {0x143f120?, 0xc000a453b0?})
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/error_strategy.go:132 +0xc6
github.com/earthly/earthly/ast/parser.(*EarthParser).Stmt.func2()
        /earthly/ast/parser/earth_parser.go:2065 +0xbb
panic({0x1130580, 0xc000a453b0})
        /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/earthly/earthly/ast/parser.(*EarthParser).Stmt(0xc0006cb5b0)
        /earthly/ast/parser/earth_parser.go:2127 +0x4f9
github.com/earthly/earthly/ast/parser.(*EarthParser).Stmts(0xc0006cb5b0)
        /earthly/ast/parser/earth_parser.go:1877 +0x1f7
github.com/earthly/earthly/ast/parser.(*EarthParser).EarthFile(0xc0006cb5b0)
        /earthly/ast/parser/earth_parser.go:767 +0x385
github.com/earthly/earthly/ast.newEarthfileTree({0xc0003180a0, 0x9}, 0xc0000ca520, {0x1449800?, 0xc0000ea200})
        /earthly/ast/ast.go:83 +0x348
github.com/earthly/earthly/ast.Parse({0x1443d58?, 0xc0000d6190}, {0xc0003180a0, 0x9}, 0x1)
        /earthly/ast/ast.go:25 +0x14c
github.com/earthly/earthly/buildcontext.(*Resolver).parseEarthfile.func1({0x1443d58?, 0xc0000d6190?}, {0x1069f20?, 0xc0002c6480?})
        /earthly/buildcontext/resolver.go:117 +0xa5
github.com/earthly/earthly/util/syncutil/synccache.(*SyncCache).Do.func1()
        /earthly/util/syncutil/synccache/synccache.go:44 +0x50
created by github.com/earthly/earthly/util/syncutil/synccache.(*SyncCache).Do
        /earthly/util/syncutil/synccache/synccache.go:40 +0x151
```

### PR DESCRIPTION
**Pull request by nelsam**
In certain edge cases, erroneous indentation was causing us to call popRecipeMode too many times. Simply adding a 'recover' changes our panic into a syntax error.

Resolves #2603

### CODE REVIEW

File: tests/invalid/Earthfile [hunk 1]
```diff
@@ -20,18 +20,24 @@ WORKDIR /test
 test-trailing-backslash:
     DO +RUN_EARTHLY_ARGS --earthfile=trailing-backslash.earth --should_fail=true --target=+base
 
+test-leading-whitespace:
+    DO +RUN_EARTHLY_ARGS --earthfile=leading-whitespace.earth --should_fail=true --target=+base --output_contains="syntax error"
```
Comment (alexcb) [inline tests/invalid/Earthfile:24]: this feels incorrect,

```
VERSION 0.6
FROM alpine
ARG foo
RUN echo foo is $foo
```

should be equivalent to 

```
VERSION 0.6
                            FROM alpine
                            ARG foo
                            RUN echo foo is $foo
```

which are both valid Earthfiles, which should output:

```
$ earthly --no-cache  +base --foo=bar
 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)

 2. Build 🔧
————————————————————————————————————————————————————————————————————————————————

              alpine | --> Load metadata alpine linux/amd64
               +base | --> FROM alpine
               +base | [----------] 100% resolve docker.io/library/alpine@sha256:f271e74b17ced29b915d351685fd4644785c6d1559dd1f2d4189a5e851ef753a
               +base | foo=bar
               +base | --> RUN echo foo is $foo
               +base | foo is bar
              output | --> exporting outputs

 3. Push ⏫ (disabled)
————————————————————————————————————————————————————————————————————————————————

To enable pushing use earthly --push

 4. Local Output 🎁
————————————————————————————————————————————————————————————————————————————————



========================== 🌍 Earthly Build  ✅ SUCCESS ==========================
```

**Review by alexcb — COMMENTED** [review]
I'm unsure about adding the test checking for syntax error; it's better than a panic, but feels wrong.

maybe instead the test could just validate no panic occurs, then we add a TODO to fix it later.

File: tests/invalid/Earthfile (hunk 1 above)
Comment (nelsam) [reply tests/invalid/Earthfile:24]: I did originally write the test to just `grep -v panic earthly.output`, but that was passing even when earthly was panicking. It felt like I was going to have to make a lot more changes to get that to work.  I'll revisit that though. There are probably other tests we'll want to use that for, so it's probably worth doing.

FWIW, since the `Earthfile` language relies on significant whitespace at the top level, I honestly think that erroneous indentation _should_ be a syntax error (again, just at the top level).

We only have one level of significant whitespace - i.e. our parser grammar only ever cares whether or not there is indentation, never how much indentation there is. Indenting more or less doesn't change anything once you've started an `INDENT`. But it makes sense that leading whitespace at the top level would trigger an `INDENT` token, and that an unexpected `INDENT` token would cause parser errors.

Comment (alexcb) [reply tests/invalid/Earthfile:24]: > I honestly think that erroneous indentation should be a syntax error (again, just at the top level).

My mistake! for some reason I thought this was valid,

```
foo:
FROM alpine
RUN echo "this will cause a syntax error since there's no leading whitespace"
```

**Review by alexcb — APPROVED** [review]
